alphabase/smiles/peptide.py

```diff
@@ -0,0 +1,196 @@
+            The SMILES string of the peptide molecule.
+        """
+        if mods and mod_site:
+            mods = {int(m): mod for m, mod in zip(mod_site.split(";"), mods.split(";"))}
+        else:
+            mods = {}
+        peptide_mol = self._build_peptide(sequence, mods)
+        return Chem.MolToSmiles(peptide_mol)
+
+    def _build_peptide(
+        self,
+        sequence: str,
+        mods: Dict[int, str],
+    ) -> Chem.Mol:
+        """
+        Build the peptide molecule from the sequence and modifications.
+
+        Parameters
+        ----------
+        sequence : str
+            Peptide sequence.
+        mods : Dict[int, str]
+            Modifications dictionary with the site as key and the modification as value.
+
+        Returns
+        -------
+        Chem.Mol
+            The peptide molecule.
+        """
+        # List to hold the amino acid molecules
+        amino_acid_mols = []
+
+        n_term_placeholder_mol = Chem.MolFromSmiles(
+            self.amino_acid_modifier.N_TERM_PLACEHOLDER, sanitize=False
+        )
+        c_term_placeholder_mol = Chem.MolFromSmiles(
+            self.amino_acid_modifier.C_TERM_PLACEHOLDER, sanitize=False
+        )
+        n_term_mod = None
+        c_term_mod = None
+
+        if 0 in mods:
+            n_term_mod = mods[0]
+        if -1 in mods:
+            c_term_mod = mods[-1]
+        # Process each amino acid in the sequence
+        for idx, aa in enumerate(sequence):
+            if idx + 1 in mods:
+                aa_smiles = self.amino_acid_modifier.ptm_dict.get(mods[idx + 1], None)
+            else:
+                aa_smiles = self.amino_acid_modifier.aa_smiles.get(aa)
+            if not aa_smiles:
+                raise ValueError(
+                    f"Unknown amino acid code: {aa} or modification: {mods.get(idx + 1, 'no mod')} (SMILES for it might be missing)"
+                )
+            # Convert the amino acid SMILES to a molecule
+            aa_mol = Chem.MolFromSmiles(aa_smiles)
+            if aa_mol is None:
+                raise ValueError(
+                    f"Invalid SMILES for amino acid {aa}, mod {mods.get(idx + 1, 'no mod')}: {aa_smiles}"
+                )
+            amino_acid_mols.append(aa_mol)
+
+        # Now, assemble the peptide
+        peptide_mol = amino_acid_mols[0]
+        peptide_mol = self.amino_acid_modifier._apply_n_terminal_modification(
+            peptide_mol,
+            n_term_placeholder_mol=n_term_placeholder_mol,
+            n_term_mod=n_term_mod,
+        )
+        for idx in range(1, len(amino_acid_mols)):
+            peptide_mol = self._connect_amino_acids(peptide_mol, amino_acid_mols[idx])
+            peptide_mol = self.amino_acid_modifier._apply_n_terminal_modification(
+                peptide_mol,
+                n_term_placeholder_mol=n_term_placeholder_mol,
+                n_term_mod=None,
+            )
+        peptide_mol = self.amino_acid_modifier._apply_c_terminal_modification(
```

Comment:
just for understanding, when are the N-term and C-term placeholders removed?
Or is this solved by concatenating them and applying a C-term water, N-Term ammonia modification?

Reply:
The N-term modification is applied to the first amino acid (what is line 102 now), after that, we start growing the peptide, and after each addition of the next amino acid we use `_apply_n_terminal_modification` to get rid of the second N-term placeholder (line 109). When we add the last AA, we modify the C-term of the resulting peptide (line 114).

Reply:
Do smiles start with N-terminal amino group of peptides?

Reply:
The string itself? Not necessarily, but we shouldn't work with it with string methods anyway. Please check out the structures in the jupyter notebook, I hope I didn't mess something up with the order or something

Reply:
Yes, for graph, it is not necessary.